olive/engine/footprint.py

```diff
@@ -0,0 +1,178 @@
+                        is_goals_met.append(v.metrics.value[metric_name] * cmp_direction >= _goal)
+                else:
+                    logger.warning(f"Metric {metric_name} is not in the objective dict")
+            self.nodes[k].metrics.is_goals_met = all(is_goals_met)
+
+    def record(self, foot_print_node: FootprintNode = None, **kwargs):
+        _model_id = kwargs.get("model_id", None)
+        if foot_print_node is not None:
+            _model_id = foot_print_node.model_id
+            self.nodes[_model_id] = foot_print_node
+        elif _model_id in self.nodes:
+            self.nodes[_model_id].update(**kwargs)
+        else:
+            self.nodes[_model_id] = FootprintNode(**kwargs)
+        self.resolve_metrics()
+
+    def get_candidates(self):
+        return {k: v for k, v in self.nodes.items() if v.metrics is not None}
+
+    def mark_pareto_frontier(self):
+        if self.is_marked_pareto_frontier:
+            return
+        for k, v in self.nodes.items():
+            # if current point's metrics is less than any other point's metrics, it is not pareto frontier
+            cmp_flag = True and v.metrics is not None and len(v.metrics.value) > 0
+            for _, _v in self.nodes.items():
+                if not cmp_flag:
+                    break
+                if _v.metrics is not None and len(_v.metrics.value) > 0:
+                    _against_pareto_frontier_check = True
+                    # if all the metrics of current point is less than any other point's metrics,
+                    # it is not pareto frontier e.g. current point's metrics is [1, 2, 3],
+                    # other point's metrics is [2, 3, 4], then current point is not pareto frontier
+                    # but if current point's metrics is [3, 2, 3], other point's metrics is [2, 3, 4],
+                    # then current point is pareto frontier
+                    for metric_name in v.metrics.value:
+                        other_point_metrics = _v.metrics.value[metric_name] * _v.metrics.cmp_direction[metric_name]
+                        current_point_metrics = v.metrics.value[metric_name] * v.metrics.cmp_direction[metric_name]
+                        _against_pareto_frontier_check &= current_point_metrics < other_point_metrics
+                    cmp_flag &= not _against_pareto_frontier_check
+            self.nodes[k].is_pareto_frontier = cmp_flag
+        self.is_marked_pareto_frontier = True
+
+    def get_pareto_frontier(self):
+        self.mark_pareto_frontier()
+        rls = {k: v for k, v in self.nodes.items() if v.is_pareto_frontier}
+        for _, v in rls.items():
+            logger.info(f"pareto frontier points: {v.model_id} {v.metrics.value}")
+
+        # restructure the pareto frontier points to instance of Footprints node for further analysis
+        return Footprint(nodes=rls, objective_dict=self.objective_dict, is_marked_pareto_frontier=True)
+
+    def _plot_pareto_frontier(self, index=[0, 1]):
```

Comment:
>index=[0, 1] [](http://example.com/codeflow?start=36&length=12)

Not used?

Reply:
I were thinking to plot pareto frontier with >2D data in the future. Then we can use index to plot any 2 or 3 dimensions.
That says:
if we take accuracy/latency/costs into account in the same time, when visualization, user can use index to chose which 2 or 3 dimension they want to display.

Reply:
I am thinking here we will hit the mutable default parameter issues?